internal/controller/storageclient_controller.go

```diff
@@ -419,9 +423,36 @@ func (r *storageClientReconcile) reconcilePhases() (ctrl.Result, error) {
 		}
 	}
 
+	if err := r.deleteStaleResources(resourceMapping); err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to delete stale resources: %v", err)
+	}
+
 	return reconcile.Result{}, nil
 }
 
+func (r *storageClientReconcile) deleteStaleResources(resourceMapping gvkToObjectKeys) error {
+	for gvk, providerSentKeys := range resourceMapping {
+		existingObjList := &metav1.PartialObjectMetadataList{}
+		existingObjList.SetGroupVersionKind(gvk)
+		if err := r.list(existingObjList); err != nil {
+			return err
+		}
+		retainKeys := make(map[types.NamespacedName]bool, len(providerSentKeys))
+		for idx := range providerSentKeys {
+			retainKeys[providerSentKeys[idx]] = true
+		}
+		for idx := range existingObjList.Items {
+			obj := &existingObjList.Items[idx]
+			if metav1.IsControlledBy(obj, &r.storageClient) && !retainKeys[client.ObjectKeyFromObject(obj)] {
+				if err := r.Delete(r.ctx, obj); client.IgnoreNotFound(err) != nil {
+					return err
```

Comment:
This logic is dangerous, you are deleting anything that you can find of a type. Please limit what you can find based on an OwnerRef (or a label if not possible)

Reply:
> you are deleting anything that you can find of a type.
- `metav1.IsControlledBy(obj, &r.storageClient)` do you mean something else?

Reply:
True, resolving